### PR TITLE
[SPARK-42088][PYTHON][WINDOWS][BUILD] Improve setup.py adaptation for Windows

### PR DESCRIPTION
**Pull request by zekai-li**
### What changes were proposed in this pull request?

Update the supports symlinks method in setup.py

### Why are the changes needed?

My system version is windows 10, and I can run setup.py with administrator permissions, so there will be no error. However, it may be troublesome for us to upgrade permissions with Windows Server, so we need to modify the code of setup.py to ensure no error. To avoid the hassle of compiling for the user, I suggest modifying the following code to enable the out-of-the-box effect

### Does this PR introduce _any_ user-facing change?

Such an update would allow users to build spark without dealing with windows specific situations

### How was this patch tested?

need to run setup.py with and without administrator permissions in windows to eventually build python packages properly

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -104,7 +105,7 @@
 
 def _supports_symlinks():
     """Check if the system supports symlinks (e.g. *nix) or not."""
-    return getattr(os, "symlink", None) is not None
+    return getattr(os, "symlink", None) is not None and ctypes.windll.shell32.IsUserAnAdmin() != 0 if sys.platform == "win32" else True
```
Comment (HyukjinKwon) [inline python/setup.py:108]: You will have to do something like this to minimize the potential breakage:

```suggestion
    return hasattr(os, "symlink") and (
        (not hasattr(ctypes, "windll"))  # Non-Windows
        or (
            # In some Windows, `os.symlink` works only for admins.
            hasattr(ctypes, "windll")
            and hasattr(ctypes.windll, "shell32")
            and hasattr(ctypes.windll.shell32, "IsUserAnAdmin")
            and ctypes.windll.shell32.IsUserAnAdmin() != 0
        )
    )
```

**Comment by HyukjinKwon** [comment]
@zekai-li mind keeping the PR template (https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE) please?

**Comment by zekai-li** [comment]
> @zekai-li mind keeping the PR template (https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE) please?

Modified, thank you very much. First attempt to submit pr to spark

**Review by HyukjinKwon — APPROVED** [review]

**Review by zekai-li — COMMENTED** [review]
nice, avoid some special problems

File: python/setup.py [hunk 2]
```diff
@@ -104,7 +105,16 @@
 
 def _supports_symlinks():
     """Check if the system supports symlinks (e.g. *nix) or not."""
-    return getattr(os, "symlink", None) is not None
+    return hasattr(os, "symlink") and (
+        (not hasattr(ctypes, "windll"))  # Non-Windows
+        or (
+            # In some Windows, `os.symlink` works only for admins.
+            hasattr(ctypes, "windll")
+            and hasattr(ctypes.windll, "shell32")
```
Comment (HyukjinKwon) [inline python/setup.py:113]:
```suggestion
            hasattr(ctypes.windll, "shell32")
```

File: python/setup.py [hunk 3]
```diff
@@ -104,7 +105,16 @@
...
+            and hasattr(ctypes.windll.shell32, "IsUserAnAdmin")
+            and ctypes.windll.shell32.IsUserAnAdmin() != 0
```
Comment (HyukjinKwon) [inline python/setup.py:115]:
```suggestion
            and bool(ctypes.windll.shell32.IsUserAnAdmin())
```

File: python/setup.py [hunk 4]
```diff
@@ -104,7 +105,16 @@
 
 def _supports_symlinks():
     """Check if the system supports symlinks (e.g. *nix) or not."""
-    return getattr(os, "symlink", None) is not None
+    return hasattr(os, "symlink") and (
+        (not hasattr(ctypes, "windll"))  # Non-Windows
+        or (
+            # In some Windows, `os.symlink` works only for admins.
```
Comment (HyukjinKwon) [inline python/setup.py:111]:
```suggestion
            # In some Windows, `os.symlink` works but only for admins.
```

**Review by HyukjinKwon — COMMENTED** [review]
Some more nits

**Review by zekai-li — COMMENTED** [review]
ok

**Review by HyukjinKwon — APPROVED** [review]

**Review by zekai-li — COMMENTED** [review]
ok

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**